src/Toolkit/PubSub/V1/PublisherApi.php

```diff
@@ -0,0 +1,673 @@
+    }
+
+    /**
+     * Parses the project from the given fully-qualified path which
+     * represents a topic resource.
+     */
+    public static function parseProjectFromTopicName($topicName)
+    {
+        return self::getTopicNameTemplate()->match($topicName)['project'];
+    }
+
+    /**
+     * Parses the topic from the given fully-qualified path which
+     * represents a topic resource.
+     */
+    public static function parseTopicFromTopicName($topicName)
+    {
+        return self::getTopicNameTemplate()->match($topicName)['topic'];
+    }
+
+    private static function getProjectNameTemplate()
+    {
+        if (self::$projectNameTemplate == null) {
+            self::$projectNameTemplate = new PathTemplate('projects/{project}');
+        }
+
+        return self::$projectNameTemplate;
+    }
+
+    private static function getTopicNameTemplate()
+    {
+        if (self::$topicNameTemplate == null) {
+            self::$topicNameTemplate = new PathTemplate('projects/{project}/topics/{topic}');
+        }
+
+        return self::$topicNameTemplate;
+    }
+
+    private static function getPageStreamingDescriptors()
+    {
+        $listTopicsPageStreamingDescriptor =
+                new PageStreamingDescriptor([
+                    'requestPageTokenField' => 'page_token',
+                    'responsePageTokenField' => 'next_page_token',
+                    'resourceField' => 'topics',
+                ]);
+        $listTopicSubscriptionsPageStreamingDescriptor =
+                new PageStreamingDescriptor([
+                    'requestPageTokenField' => 'page_token',
+                    'responsePageTokenField' => 'next_page_token',
+                    'resourceField' => 'subscriptions',
+                ]);
+
+        $pageStreamingDescriptors = [
+            'listTopics' => $listTopicsPageStreamingDescriptor,
+            'listTopicSubscriptions' => $listTopicSubscriptionsPageStreamingDescriptor,
+        ];
+
+        return $pageStreamingDescriptors;
+    }
+
+    // TODO(garrettjones): add channel (when supported in gRPC)
+    /**
+     * Constructor.
+     *
+     * @param array $options {
+     *                       Optional. Options for configuring the service API wrapper.
+     *
+     *     @var string $serviceAddress The domain name of the API remote host.
+     *                                  Default 'pubsub.googleapis.com'.
+     *     @var mixed $port The port on which to connect to the remote host. Default 443.
+     *     @var Grpc\ChannelCredentials $sslCreds
+     *           A `ChannelCredentials` for use with an SSL-enabled channel.
+     *           Default: a credentials object returned from
+     *           Grpc\ChannelCredentials::createSsl()
+     *     @var array $scopes A string array of scopes to use when acquiring credentials.
+     *                         Default the scopes for the Google Cloud Pub/Sub API.
+     *     @var array $retryingOverride
```

Comment:
Is there precedence for allowing configuration of retries per method at the constructor level and at the individual method level? I haven't seen this before.